File: Face.java
package ModelLib;

import java.util.Vector;

import static ModelLib.Utils.cross;
import static ModelLib.Utils.normalize;

public class Face
{
    protected int[] vids = new int[3];
    protected Vector<Neighbor> neighbors;
    protected Point normal;
    protected Point c;
    protected int label;

    protected Face()
    {
        label = 0;
    }

    protected Face(Vertex v1, Vertex v2, Vertex v3)
    {
        c = (v1.p.sum(v2.p.sum(v3.p))).div(3.0);
        normal = cross(v2.p.sub(v1.p), v3.p.sub(v1.p));
        normalize(normal);
        //if (dot(normal, c) < 0) normal = normal * (-1.0);
        label = 0;
    }

    protected Face(Face f)
    {
        vids = f.vids;
        neighbors = f.neighbors;
        normal = f.normal;
        c = f.c;
        label = f.label;
    }
}

File: src/Face.cpp
//
// Created by timofey on 28.09.2021.
//
#include "../include/Face.h"

using namespace util;

Face::Face()
{
    label = 0;
}

Face::Face(Vertex& v1, Vertex& v2, Vertex& v3)
{
    c = (v1.p + v2.p + v3.p) / 3.0;
    normal = cross(v2.p - v1.p, v3.p - v1.p);
    normalizeV(normal);
    label = 0;
}
Face::Face(const Face& f)
{
    vids[0] = f.vids[0];
    vids[1] = f.vids[1];
    vids[2] = f.vids[2];
    neighs = f.neighs;
    normal = f.normal;
    c = f.c;
    label = f.label;
}